[FIX] sale_coupon_advanced: clean up forced lines

oerp-odoo opened the pull request:
Applying bandage fix to clean up incorrectly created forced sale order
lines.

Bugs fixed:
    - no longer adds unrelated programs reward products on SO when
        applying coupon code from forced program.
    - no longer adds products from forced promotions that need code to
        be entered, when recomputing promotions.
    - handling order line sequence generation, when it is called, when
        no lines exists.

oerp-odoo committed:
[FIX] sale_coupon_advanced: clean up forced lines

Applying bandage fix to clean up incorrectly created forced sale order
lines.

Bugs fixed:
    - no longer adds unrelated programs reward products on SO when
        applying coupon code from forced program.
    - no longer adds products from forced promotions that need code to
        be entered, when recomputing promotions.
    - handling order line sequence generation, when it is called, when
        no lines exists.

i-vyshnevska commented on sale_coupon_advanced/models/sale_order.py:
```diff
@@ -6,15 +6,30 @@
         """Add the rewarded product if a reward line has been found for this
         product
         """
+        # NOTE. This method dosctring is misleading, it says, it adds
```
you can update docstring with the real work method doing

oerp-odoo replied:
I thought first to highlight it for review. But you are right, I guess I can just update myself.

i-vyshnevska commented on sale_coupon_advanced/models/sale_coupon_program.py:
```diff
@@ -66,6 +69,8 @@ def _filter_programs_from_common_rules(self, order, next_order=False):
+    # TODO: method docstring does not make sense. It does not do
+    # anything what is described.
     def _force_sale_order_lines(self, programs, order):
         """Return the programs when `is_reward_product_forced` is selected
```
same, you are free to refactor docstring and code as needed

i-vyshnevska commented on sale_coupon_advanced/models/sale_order.py:
```diff
@@ -104,10 +119,38 @@ def _remove_invalid_reward_lines(self):
         # TODO: rollback forced lines which is not used for creation of
         # reward lines for other programs
```
```suggestion
```

oerp-odoo committed:
[FIX] sale_coupon_advanced: counter forced reward

Previous fix was not enough after odoo source code was updated, which
still caused to break functionality.

In order for it to work in all cases, instead of removing incorrectly
forced lines, now we check if line can be added and add that line before
it is expected to be used by sale_coupon module.

Because trigger points are different: for applying coupon and
recomputing promotions, had to inject this logic into multiple places.
